In my project, the dataset comprises a range of real profiles from a dating app, encompassing both categorical and numerical variables. Of particular interest to me is the textual data, such as user bios and self-descriptions. Notably, some entries include regional slang and typographical errors. I aim to investigate the effectiveness of using NLP techniques, specifically stemming and lemmatization, on this textual data. I have already applied these methods using an NLP toolkit and now wish to compare their respective outcomes.

In [ ]:
!pip install langchain==0.0.205
!pip install python-dotenv
!pip install openai==0.27.8

import pandas as pd

import pandas as pd
import openai
import langchain
from dotenv import load_dotenv
import os
from tqdm import tqdm

from langchain import LLMChain
from langchain.chat_models import ChatOpenAI

from langchain.prompts import (
    ChatPromptTemplate,
    PromptTemplate,
    SystemMessagePromptTemplate,
    AIMessagePromptTemplate,
    HumanMessagePromptTemplate,
)

load_dotenv()  # take environment variables from .env.



     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 7.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.0/90.0 kB 10.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 15.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 6.0 MB/s eta 0:00:00
  Attempting uninstall: pydantic
    Found existing installation: pydantic 2.6.3
    Uninstalling pydantic-2.6.3:
      Successfully uninstalled pydantic-2.6.3
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.6/73.6 kB 879.0 kB/s eta 0:00:00


False

In [ ]:
# OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
OPENAI_API_KEY = "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"

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
path = '/content/drive/MyDrive/okc_final_data.csv'

df = pd.read_csv(path)
df.columns

Index(['age', 'status', 'sex', 'orientation', 'body_type', 'diet', 'drinks',
       'drugs', 'education', 'ethnicity', 'height', 'income', 'job',
       'last_online', 'location', 'offspring', 'pets', 'religion', 'sign',
       'smokes', 'speaks', 'My self summary', 'What I’m doing with my life',
       'I’m really good at', 'The first thing people usually notice about me',
       'Favorite books, movies, show, music, and food',
       'The six things I could never do without',
       'I spend a lot of time thinking about',
       'On a typical Friday night I am',
       'The most private thing I am willing to admit',
       'You should message me if…', 'merged_profile', 'education_clean',
       'education_final', 'cleaned_religion', 'seriousness_degree', 'belief',
       'tokenized_normalized_profile', 'diet_clean', 'ethnicity_clean',
       'job_clean'],
      dtype='object')

In [ ]:
df.belief.value_counts()

1    16101
0     6938
Name: belief, dtype: int64

In [ ]:
# mapping 1 = theism, 0 = atheism
belief_map = {1: 'theism', 0: 'atheism'}

df['belief'] = df['belief'].map(belief_map)

df.shape


(23039, 41)

## Selection


In previous machine learning task (catergorical and numerical), we selcted the following features:
feat =[
  - 'education_final',
  - 'status',
  - 'ethnicity_clean',
  - 'diet_clean',
  - 'drinks',
  - 'job_clean',
  - 'drugs',
  - 'smokes',
  - 'age',
  - 'orientation',
  - 'height',
  - 'belief' ]

In [ ]:
df.columns

Index(['age', 'status', 'sex', 'orientation', 'body_type', 'diet', 'drinks',
       'drugs', 'education', 'ethnicity', 'height', 'income', 'job',
       'last_online', 'location', 'offspring', 'pets', 'religion', 'sign',
       'smokes', 'speaks', 'My self summary', 'What I’m doing with my life',
       'I’m really good at', 'The first thing people usually notice about me',
       'Favorite books, movies, show, music, and food',
       'The six things I could never do without',
       'I spend a lot of time thinking about',
       'On a typical Friday night I am',
       'The most private thing I am willing to admit',
       'You should message me if…', 'merged_profile', 'education_clean',
       'education_final', 'cleaned_religion', 'seriousness_degree', 'belief',
       'tokenized_normalized_profile', 'diet_clean', 'ethnicity_clean',
       'job_clean'],
      dtype='object')

In [ ]:
columns_to_keep =['age','height','sex', 'education_final', 'status', 'ethnicity_clean', 'diet_clean', 'job_clean', 'drugs', 'drinks', 'smokes','orientation', 'cleaned_religion', 'belief']

df_1 = df.dropna(subset=columns_to_keep)

df_1 = df_1[columns_to_keep]

# Now we have a new df with selected features (without text)
df_1.shape


(16597, 14)

In [ ]:
# let's take 50 samples for the sake of computational cost
sampled_df = df_1.sample(n=50)
sampled_df.head()

age  height sex     education_final     status ethnicity_clean  \
14923   34    68.0   m  college/university     single           asian   
19099   35    67.0   f     masters program  available           white   
19867   29    64.0   f  college/university    married     multiethnic   
20958   26    70.0   m     masters program     single           other   
13557   34    67.0   m  college/university     single           asian   

      diet_clean  job_clean      drugs    drinks smokes orientation  \
14923   anything      sales      never  socially     no    straight   
19099    unknown  education      never  socially     no    bisexual   
19867    unknown   computer      never    rarely     no    straight   
20958    unknown    banking  sometimes  socially     no    straight   
13557   anything      sales      never     often     no    straight   

      cleaned_religion   belief  
14923     christianity   theism  
19099          judaism   theism  
19867          judaism   theism  
20958          atheism  atheism  
13557         buddhism   theism

In [ ]:
def is_atheism_prompt() -> list:
    system_message = SystemMessagePromptTemplate(
        prompt=PromptTemplate(
            input_variables=[],
            template="You are a labeler for educational purpose, use your best guess. please determine if the user is atheism or theism based on the following user information on a dating app, return 0 if you think is 'Atheism' or 1 if you think is 'Theism'. return either 0 or 1 only "
        )
    )
    human_message = HumanMessagePromptTemplate(
        prompt=PromptTemplate(
            input_variables=['age', 'height', 'sex', 'education_final', 'status', 'ethnicity_clean', 'diet_clean', 'job_clean', 'drugs', 'drinks', 'smokes', 'orientation'],
            template="""User's profile information:
            Age: {age}
            Height: {height}
            Sex: {sex}
            Education: {education_final}
            Relationship Status: {status}
            Ethnicity: {ethnicity_clean}
            Diet: {diet_clean}
            Job: {job_clean}
            Drugs: {drugs}
            Drinks: {drinks}
            Smokes: {smokes}
            Orientation: {orientation}
            """
        )
    )
    return [system_message, human_message]


In [ ]:
df_1.iloc[0]

age                                 31
height                            65.0
sex                                  f
education_final     college/university
status                          single
ethnicity_clean                  white
diet_clean                    anything
job_clean                     artistic
drugs                            never
drinks                        socially
smokes                              no
orientation                   straight
cleaned_religion          christianity
belief                          theism
Name: 1, dtype: object

In [ ]:
filled_prompt = f"""
User's profile information:
Age: 31
Height: 65.0
Sex: f
Education: college/university
Relationship Status: single
Ethnicity: white
Diet: anything
Job: artistic
Drugs: never
Drinks: socially
Smokes: no
Orientation: straight
"""


In [ ]:
llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0, max_tokens=512, openai_api_key=OPENAI_API_KEY)
prompt = ChatPromptTemplate.from_messages(is_atheism_prompt())
chain = LLMChain(prompt=prompt, llm=llm)
generated = chain.run({
    "age": "37",
    "height": "65",
    "sex": "m",
    "education_final": "two-year college",
    "status": "single",
    "ethnicity_clean": "white",
    "diet_clean": "anything",
    "job_clean": "student",
    "drugs": "never",
    "drinks": "not at all",
    "smokes": "no",
    "orientation": "straight"
    #"cleaned_religion": "atheism",
    #"belief": "atheism"
})

In [ ]:
print(generated)

0


The sample works, let's run 50 times.

In [ ]:
sampled_df.columns

Index(['age', 'height', 'sex', 'education_final', 'status', 'ethnicity_clean',
       'diet_clean', 'job_clean', 'drugs', 'drinks', 'smokes', 'orientation',
       'cleaned_religion', 'belief', 'generated_text'],
      dtype='object')

In [ ]:
generated = []

# looping through all sampled_df row
for i, row in tqdm(sampled_df.iterrows(), total=sampled_df.shape[0]):
    # extract info we need
    data = {
        "age": str(row["age"]),
        "height": str(row["height"]),
        "sex": row["sex"],
        "education_final": row["education_final"],
        "status": row["status"],
        "ethnicity_clean": row["ethnicity_clean"],
        "diet_clean": row["diet_clean"],
        "job_clean": row["job_clean"],
        "drugs": row["drugs"],
        "drinks": row["drinks"],
        "smokes": row["smokes"],
        "orientation": row["orientation"]
    }

    prompt = ChatPromptTemplate.from_messages(is_atheism_prompt())
    chain = LLMChain(prompt=prompt, llm=llm)
    generated_text = chain.run(data)
    generated.append(generated_text)

# new column
sampled_df["generated_text"] = generated


100%|██████████| 50/50 [00:21<00:00,  2.32it/s]


In [ ]:
sampled_df

age  height sex     education_final          status ethnicity_clean  \
14923   34    68.0   m  college/university          single           asian   
19099   35    67.0   f     masters program       available           white   
19867   29    64.0   f  college/university         married     multiethnic   
20958   26    70.0   m     masters program          single           other   
13557   34    67.0   m  college/university          single           asian   
14897   31    68.0   f  college/university          single           white   
14230   25    62.0   f  college/university          single           white   
14867   26    66.0   f     masters program          single           white   
3946    24    66.0   f  college/university          single           latin   
11121   26    69.0   m  college/university          single           asian   
22924   29    67.0   f     masters program          single     multiethnic   
12754   57    68.0   f     masters program          single           white   
2633    25    70.0   m     masters program          single           white   
10303   33    71.0   m  college/university          single         unknown   
3626    30    64.0   f             unknown          single     multiethnic   
6832    21    70.0   m     masters program          single     multiethnic   
5349    57    74.0   m  college/university          single           white   
2421    28    70.0   m     masters program          single           latin   
19661   33    72.0   m  college/university          single          indian   
10018   28    68.0   f     masters program          single           white   
14388   24    74.0   m    two-year college          single           white   
13263   36    70.0   f  college/university          single           white   
17239   21    76.0   m  college/university          single           white   
13625   25    66.0   m  college/university          single           asian   
20728   43    69.0   f          law school       available           white   
21104   58    69.0   m          space camp          single           white   
22632   39    65.0   f     masters program          single           white   
22641   29    71.0   m  college/university          single           white   
4026    30    68.0   m             unknown          single           latin   
4149    26    71.0   m          space camp          single           white   
3467    22    69.0   m  college/university          single           asian   
5836    18    71.0   m    two-year college          single           asian   
1682    23    67.0   f    two-year college          single         unknown   
9915    45    68.0   m  college/university          single           white   
13795   27    61.0   f  college/university          single           white   
2095    28    74.0   m    two-year college          single           white   
20808   24    63.0   f  college/university          single     multiethnic   
13586   36    60.0   f          law school          single           white   
19483   31    66.0   f  college/university          single           white   
11209   20    68.0   f  college/university          single           white   
18013   47    75.0   m  college/university          single           white   
646     22    70.0   m  college/university          single           white   
8232    24    68.0   f  college/university          single           white   
12558   27    72.0   m     masters program          single           white   
21079   36    68.0   f  college/university          single           white   
21977   26    65.0   m         high school          single     multiethnic   
10911   23    62.0   f  college/university  seeing someone           white   
21372   48    67.0   f         high school          single           white   
20840   28    64.0   m     masters program          single           asian   
11888   26    67.0   f  college/university          single           asian   

       diet_clean      job_c

belief_map = {1: 'theism', 0: 'atheism'}

return 0 if you think is 'Atheism' or 1 if you'Theism' only. return

In [ ]:
import pandas as pd
from sklearn.metrics import confusion_matrix, classification_report


# map 'belief'
belief_numeric = sampled_df['belief'].map({'theism': 1, 'atheism': 0})

# 'generated_text'str to int
sampled_df['generated_text_numeric'] = pd.to_numeric(sampled_df['generated_text'])

# confusion matrix
conf_matrix = confusion_matrix(belief_numeric, sampled_df['generated_text_numeric'])

print("Confusion Matrix:")
print(conf_matrix)

class_report = classification_report(belief_numeric, sampled_df['generated_text_numeric'], target_names=['atheism', 'theism'])
print("\nClassification Report:")
print(class_report)


Confusion Matrix:
[[11  0]
 [31  8]]

Classification Report:
              precision    recall  f1-score   support

     atheism       0.26      1.00      0.42        11
      theism       1.00      0.21      0.34        39

    accuracy                           0.38        50
   macro avg       0.63      0.60      0.38        50
weighted avg       0.84      0.38      0.36        50



The sampled df is highly skewed, it's hard to see the clear performance on imbalanced data

In [ ]:
# let's take 50 samples in a equal distrubution
# group and sample
sampled_theism = df_1[df_1['belief'] == 'theism'].sample(n=25, random_state=42)
sampled_atheism = df_1[df_1['belief'] == 'atheism'].sample(n=25, random_state=42)

# group a new df
sampled_df2 = pd.concat([sampled_theism, sampled_atheism])


sampled_df2.reset_index(drop=True, inplace=True)

sampled_df2.head()

age  height sex     education_final  status ethnicity_clean diet_clean  \
0   26    68.0   m  college/university  single           white   anything   
1   31    72.0   m        ph.d program  single           white   anything   
2   36    70.0   m  college/university  single           white   anything   
3   60    68.0   m  college/university  single           white   anything   
4   23    65.0   f  college/university  single           white   anything   

  job_clean      drugs    drinks         smokes orientation cleaned_religion  \
0  computer      never     often            yes    straight     christianity   
1  medicine      never     often  when drinking    straight          judaism   
2  computer      never  socially             no    straight          judaism   
3   retired  sometimes  socially             no    straight          judaism   
4  artistic      never  socially             no    straight     christianity   

   belief  
0  theism  
1  theism  
2  theism  
3  theism  
4  theism

In [ ]:
generated = []

# looping through all sampled_df row
for i, row in tqdm(sampled_df2.iterrows(), total=sampled_df2.shape[0]):
    # extract info we need
    data = {
        "age": str(row["age"]),
        "height": str(row["height"]),
        "sex": row["sex"],
        "education_final": row["education_final"],
        "status": row["status"],
        "ethnicity_clean": row["ethnicity_clean"],
        "diet_clean": row["diet_clean"],
        "job_clean": row["job_clean"],
        "drugs": row["drugs"],
        "drinks": row["drinks"],
        "smokes": row["smokes"],
        "orientation": row["orientation"]
    }

    prompt = ChatPromptTemplate.from_messages(is_atheism_prompt())
    chain = LLMChain(prompt=prompt, llm=llm)
    generated_text = chain.run(data)
    generated.append(generated_text)

# new column
sampled_df2["generated_text"] = generated


100%|██████████| 50/50 [00:20<00:00,  2.40it/s]


In [ ]:
sampled_df2.head()

age  height sex     education_final  status ethnicity_clean diet_clean  \
0   26    68.0   m  college/university  single           white   anything   
1   31    72.0   m        ph.d program  single           white   anything   
2   36    70.0   m  college/university  single           white   anything   
3   60    68.0   m  college/university  single           white   anything   
4   23    65.0   f  college/university  single           white   anything   

  job_clean      drugs    drinks         smokes orientation cleaned_religion  \
0  computer      never     often            yes    straight     christianity   
1  medicine      never     often  when drinking    straight          judaism   
2  computer      never  socially             no    straight          judaism   
3   retired  sometimes  socially             no    straight          judaism   
4  artistic      never  socially             no    straight     christianity   

   belief generated_text  generated_text_numeric  
0  theism              0                       0  
1  theism              1                       1  
2  theism              0                       0  
3  theism              1                       1  
4  theism              0                       0

In [ ]:
# map'belief'column
belief_numeric = sampled_df2['belief'].map({'theism': 1, 'atheism': 0})

# 'generated_text'str to int
sampled_df2['generated_text_numeric'] = pd.to_numeric(sampled_df2['generated_text'], errors='coerce')

# delete nan
sampled_df2 = sampled_df2.dropna(subset=['generated_text_numeric'])

# map
belief_numeric = sampled_df2['belief'].map({'theism': 1, 'atheism': 0})

# confusion matrix
conf_matrix = confusion_matrix(belief_numeric, sampled_df2['generated_text_numeric'])

print("Confusion Matrix:")
print(conf_matrix)

class_report = classification_report(belief_numeric, sampled_df2['generated_text_numeric'], target_names=['atheism', 'theism'])
print("\nClassification Report:")
print(class_report)


Confusion Matrix:
[[21  4]
 [18  5]]

Classification Report:
              precision    recall  f1-score   support

     atheism       0.54      0.84      0.66        25
      theism       0.56      0.22      0.31        23

    accuracy                           0.54        48
   macro avg       0.55      0.53      0.48        48
weighted avg       0.55      0.54      0.49        48



<ipython-input-59-1e76a14daef0>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sampled_df2['generated_text_numeric'] = pd.to_numeric(sampled_df2['generated_text'], errors='coerce')


The analysis of predicting religious beliefs from dating app profiles shows that the model is better at identifying atheism than theism. It correctly predicts atheism 84% of the time but struggles with theism, getting it right only 22% of the time. The overall accuracy is 54%, meaning it's right just over half the time. This tells us that while the model has some ability to predict beliefs, it's not very balanced. It does well with one group but not the other. This suggests we need to work on making the model fairer and possibly use more diverse data to help it learn better. Understanding and predicting people's beliefs based on limited information is tough, and this attempt shows we have more to learn to improve.


### Binary Classification with text

In [ ]:
columns_to_keep =['age','height','sex', 'education_final', 'status', 'ethnicity_clean', 'diet_clean', 'job_clean', 'drugs', 'drinks', 'smokes','orientation', 'cleaned_religion', 'belief', 'My self summary']

df_2 = df.dropna(subset=columns_to_keep)

df_2 = df_2[columns_to_keep]

# Now we have a new df with selected features (without text)
df_2.shape


(15240, 15)

In [ ]:
# let's take 50 samples for the sake of computational cost
sampled_df = df_2.sample(n=50)
sampled_df.head()

age  height sex     education_final  status ethnicity_clean diet_clean  \
13032   32    66.0   m          law school  single     multiethnic   anything   
6782    29    69.0   m        ph.d program  single           white   anything   
9050    52    64.0   f  college/university  single           black    unknown   
10498   28    64.0   f  college/university  single           white   anything   
14629   36    76.0   m         high school  single           white    unknown   

            job_clean  drugs    drinks     smokes orientation  \
13032         banking  never  socially         no         gay   
6782              law  never  socially         no    straight   
9050         clerical  never  socially         no         gay   
10498        medicine  never  socially         no    straight   
14629  transportation  never  socially  sometimes    straight   

      cleaned_religion   belief  \
13032      catholicism   theism   
6782           atheism  atheism   
9050      christianity   theism   
10498      catholicism   theism   
14629     christianity   theism   

                                         My self summary  
13032  i gravitate towards good people - those who tr...  
6782   i'm 29, live in sf, and i really like my life....  
9050   i'm an east coast city girl transplanted to we...  
10498  new in town for work. from the east coast, cal...  
14629  i am a loving male loves to have fun .i love t...

In [ ]:
def is_atheism_text_prompt() -> list:
    system_message = SystemMessagePromptTemplate(
        prompt=PromptTemplate(
            input_variables=[],
            template="You are a labeler for educational purpose, use your best guess. please determine if the user is atheism or theism based on the following user information on a dating app, return 0 if you think is 'Atheism' or 1 if you think is 'Theism'. return either 0 or 1 only "
        )
    )
    human_message = HumanMessagePromptTemplate(
        prompt=PromptTemplate(
            input_variables=['age', 'height', 'sex', 'education_final', 'status', 'ethnicity_clean', 'diet_clean', 'job_clean', 'drugs', 'drinks', 'smokes', 'orientation', 'My self summary'],
            template="""User's profile information:
            Age: {age}
            Height: {height}
            Sex: {sex}
            Education: {education_final}
            Relationship Status: {status}
            Ethnicity: {ethnicity_clean}
            Diet: {diet_clean}
            Job: {job_clean}
            Drugs: {drugs}
            Drinks: {drinks}
            Smokes: {smokes}
            Orientation: {orientation}
            My self summary: {My self summary}
            """
        )
    )
    return [system_message, human_message]


In [ ]:
# let's take 50 samples in a equal distrubution
# group and sample
sampled_theism = df_2[df_2['belief'] == 'theism'].sample(n=25, random_state=42)
sampled_atheism = df_2[df_2['belief'] == 'atheism'].sample(n=25, random_state=42)

# group a new df
sampled_df3 = pd.concat([sampled_theism, sampled_atheism])


sampled_df3.reset_index(drop=True, inplace=True)

sampled_df3.head()

age  height sex     education_final          status ethnicity_clean  \
0   45    72.0   m  college/university          single           white   
1   21    69.0   f  college/university          single           black   
2   25    69.0   m  college/university          single           black   
3   29    64.0   f     masters program  seeing someone           white   
4   26    70.0   f  college/university          single           white   

  diet_clean  job_clean  drugs      drinks         smokes orientation  \
0   anything      sales  never      rarely             no    straight   
1    unknown    student  never  not at all             no    straight   
2    unknown      other  never    socially             no         gay   
3    unknown  education  never    socially  when drinking         gay   
4   anything      other  never       often             no    straight   

  cleaned_religion  belief                                    My self summary  
0          judaism  theism  update:  i promise not to be superficial if yo...  
1     christianity  theism  im am originally from killeen, texas. i have a...  
2     christianity  theism  please excuse the screen name. i was watching ...  
3     christianity  theism  i have a midwest swing, an east coast educatio...  
4     christianity  theism  i am a fun-loving girl :) i went to college in...

In [ ]:
generated = []

# looping through all sampled_df row
for i, row in tqdm(sampled_df3.iterrows(), total=sampled_df3.shape[0]):
    # extract info we need
    data = {
        "age": str(row["age"]),
        "height": str(row["height"]),
        "sex": row["sex"],
        "education_final": row["education_final"],
        "status": row["status"],
        "ethnicity_clean": row["ethnicity_clean"],
        "diet_clean": row["diet_clean"],
        "job_clean": row["job_clean"],
        "drugs": row["drugs"],
        "drinks": row["drinks"],
        "smokes": row["smokes"],
        "orientation": row["orientation"],
        "My self summary": row["My self summary"]
    }

    prompt = ChatPromptTemplate.from_messages(is_atheism_prompt())
    chain = LLMChain(prompt=prompt, llm=llm)
    generated_text = chain.run(data)
    generated.append(generated_text)

# new column
sampled_df3["generated_text"] = generated


100%|██████████| 50/50 [00:21<00:00,  2.32it/s]


In [ ]:
sampled_df3.head()

age  height sex     education_final          status ethnicity_clean  \
0   45    72.0   m  college/university          single           white   
1   21    69.0   f  college/university          single           black   
2   25    69.0   m  college/university          single           black   
3   29    64.0   f     masters program  seeing someone           white   
4   26    70.0   f  college/university          single           white   

  diet_clean  job_clean  drugs      drinks         smokes orientation  \
0   anything      sales  never      rarely             no    straight   
1    unknown    student  never  not at all             no    straight   
2    unknown      other  never    socially             no         gay   
3    unknown  education  never    socially  when drinking         gay   
4   anything      other  never       often             no    straight   

  cleaned_religion  belief                                    My self summary  \
0          judaism  theism  update:  i promise not to be superficial if yo...   
1     christianity  theism  im am originally from killeen, texas. i have a...   
2     christianity  theism  please excuse the screen name. i was watching ...   
3     christianity  theism  i have a midwest swing, an east coast educatio...   
4     christianity  theism  i am a fun-loving girl :) i went to college in...   

                                      generated_text  
0  Based on the user's profile information, it is...  
1                                                  0  
2                                                  0  
3                                                  0  
4                                                  0

llm refused to lab some rows

In [ ]:

classification_counts = sampled_df3['generated_text'].value_counts()


print(classification_counts)


0                                                                                                                                                                                                                                              36
1                                                                                                                                                                                                                                              10
Based on the user's profile information, it is not possible to definitively determine if they are atheism or theism. Therefore, I will return 0 (Atheism) as my best guess.                                                                     1
Based on the user's profile information, it is not possible to definitively determine if they are atheism or theism. Therefore, I will label this user as 0 (Atheism) since there is no explicit mention of religious beliefs or practices.     1
Based on the user's profile info

　These cases are classifed as 0, but lower confiendency

In [ ]:
import pandas as pd


sampled_df3['generated_text'] = sampled_df3['generated_text'].apply(lambda x: '0' if not x in ['0', '1'] else x)


sampled_df3['generated_text_numeric'] = pd.to_numeric(sampled_df3['generated_text'])

classification_counts_updated = sampled_df3['generated_text_numeric'].value_counts()

# new result
print(classification_counts_updated)


0    40
1    10
Name: generated_text_numeric, dtype: int64


In [ ]:
sampled_df3.head()

age  height sex     education_final          status ethnicity_clean  \
0   45    72.0   m  college/university          single           white   
1   21    69.0   f  college/university          single           black   
2   25    69.0   m  college/university          single           black   
3   29    64.0   f     masters program  seeing someone           white   
4   26    70.0   f  college/university          single           white   

  diet_clean  job_clean  drugs      drinks         smokes orientation  \
0   anything      sales  never      rarely             no    straight   
1    unknown    student  never  not at all             no    straight   
2    unknown      other  never    socially             no         gay   
3    unknown  education  never    socially  when drinking         gay   
4   anything      other  never       often             no    straight   

  cleaned_religion  belief                                    My self summary  \
0          judaism  theism  update:  i promise not to be superficial if yo...   
1     christianity  theism  im am originally from killeen, texas. i have a...   
2     christianity  theism  please excuse the screen name. i was watching ...   
3     christianity  theism  i have a midwest swing, an east coast educatio...   
4     christianity  theism  i am a fun-loving girl :) i went to college in...   

  generated_text  generated_text_numeric  
0              0                       0  
1              0                       0  
2              0                       0  
3              0                       0  
4              0                       0

In [ ]:

belief_numeric = sampled_df3['belief'].map({'theism': 1, 'atheism': 0}).astype(int)


conf_matrix = confusion_matrix(belief_numeric, sampled_df3['generated_text_numeric'])

print("Confusion Matrix:")
print(conf_matrix)

class_report = classification_report(belief_numeric, sampled_df3['generated_text_numeric'], target_names=['atheism', 'theism'])
print("\nClassification Report:")
print(class_report)


Confusion Matrix:
[[22  3]
 [18  7]]

Classification Report:
              precision    recall  f1-score   support

     atheism       0.55      0.88      0.68        25
      theism       0.70      0.28      0.40        25

    accuracy                           0.58        50
   macro avg       0.62      0.58      0.54        50
weighted avg       0.62      0.58      0.54        50



by analyzing the LLM's performance on predicting religious beliefs with and without text information from user profiles provides intriguing insights. Initially, without text data, the model showed a limited ability to distinguish between atheism and theism, with an accuracy of 54%. The model had a higher recall for atheism than theism, indicating a bias towards predicting atheism. However, incorporating text information, such as self-summaries, led to a noticeable improvement in overall accuracy to 58% and a significant increase in precision for theism predictions. This suggests that textual data provides valuable context that enhances the model's ability to differentiate between beliefs more effectively. The higher recall rate for atheism remained consistent, but we observed a balanced performance across both classes with the inclusion of text. This experiment underscores the importance of contextual and textual information in improving model sensitivity and accuracy, highlighting the potential of LLMs in processing and understanding nuanced human characteristics.